### PR TITLE
Introduce user authentication

### PR DESCRIPTION
**Pull request by MeenaAlfons**
## Description

- Add `AuthenticateUser`
- Add `AuthorizeChannel`
- Mark `Authenticate` as deprecated and replace it with `AuthorizeChannel`

## CHANGELOG

* [ADDED] AuthenticateUser
* [ADDED] AuthorizeChannel
* [DEPRECATED] Authenticate
* [CHANGED] Bump Newtonsoft.Json to version 13.0.2

### CODE REVIEW

File: PusherServer.Tests/UnitTests/AuthenticateUser.cs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+        [OneTimeSetUp]
+        public void Setup()
+        {
+            _pusher = new Pusher(Config.AppId, Config.AppKey, Config.AppSecret);
+        }
+
+        [Test]
+        [ExpectedException(typeof(ArgumentNullException))]
+        public void null_presence_data_throw_Exception()
```
Comment (agatav) [inline PusherServer.Tests/UnitTests/AuthenticateUser.cs:22]: nit `null_user_data`

File: PusherServer.Tests/UnitTests/AuthenticateUser.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+            string expectedAuthString = Config.AppKey + ":" + CreateSignedString(socketId, userDataJson);
+
+            IUserAuthenticationResponse result = _pusher.AuthenticateUser(socketId, userData);
+            Assert.AreEqual(expectedAuthString, result.auth);
+            Assert.AreEqual(userDataJson, result.user_data);
+        }
+
+        [Test]
+        public void with_userinfo_and_watchlist_the_auth_response_is_valid()
```
Comment (agatav) [inline PusherServer.Tests/UnitTests/AuthenticateUser.cs:87]: the same test case as `with_watchlist_the_auth_response_is_valid()`

File: PusherServer.Tests/UnitTests/AuthenticateUser.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+            string expectedAuthString = Config.AppKey + ":" + CreateSignedString(socketId, userDataJson);
+
+            IUserAuthenticationResponse result = _pusher.AuthenticateUser(socketId, userData);
+            Assert.AreEqual(expectedAuthString, result.auth);
+            Assert.AreEqual(userDataJson, result.user_data);
+        }
+
+        [Test]
+        public void channel_data_is_encoded_as_JSON()
```
Comment (agatav) [inline PusherServer.Tests/UnitTests/AuthenticateUser.cs:107]: redundant test case here

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,23 +1,28 @@
 # Changelog
 
+## 4.7.0
+* [ADDED] AuthenticateUser
+* [ADDED] AuthorizeChannel
+* [DEPRECATED] Authenticate
+
```
Comment (agatav) [inline CHANGELOG.md:7]: add that you bumped `Newtonsoft.Json` version

**Review by agatav — APPROVED** [review]

File: PusherServer/UserAuthenticationResponse.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        [DataMember(Name = "user_data", IsRequired = false, EmitDefaultValue = false)]
+        public string user_data
+        {
+            get
+            {
+                return DefaultSerializer.Default.Serialize(_userData);
+            }
+        }
+
+        public string ToJson()
```
[the remark this answers is not in the document]
Comment (MeenaAlfons) [reply PusherServer/UserAuthenticationResponse.cs:52]: That's how it was for the old code. `ToString()` is made to return JSON but we provide `ToJson()` to make it more intuitive.

File: PusherServer.Tests/UnitTests/AuthenticateUser.cs (hunk 2 above)
Comment (MeenaAlfons) [reply PusherServer.Tests/UnitTests/AuthenticateUser.cs:87]: I know, setting different TestFixture for the same test is possible but `user_info` uses anonymous type which doesn't work in that context out of the box. Therefore, I left the test like this for now.

File: PusherServer.Tests/UnitTests/AuthenticateUser.cs (hunk 3 above)
Comment (MeenaAlfons) [reply PusherServer.Tests/UnitTests/AuthenticateUser.cs:107]: I agree it is redundant. I'll remove this one.

File: CHANGELOG.md (hunk 4 above)
Comment (MeenaAlfons) [reply CHANGELOG.md:7]: Sure

**Pull request merged**